sdks/java/core/src/main/java/org/apache/beam/sdk/io/WholeFileIO.java

```diff
@@ -0,0 +1,347 @@
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<KV<String, byte[]>>, PDone> {
+    private static final Logger LOG = LoggerFactory.getLogger(Write.class);
+
+    @Nullable abstract ValueProvider<ResourceId> getOutputDir();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setOutputDir(ValueProvider<ResourceId> outputDir);
+
+      abstract Write build();
+    }
+
+    public Write to(String outputDir) {
+      return to(FileSystems.matchNewResource(outputDir, true));
+    }
+
+    public Write to(ResourceId outputDir) {
+      return toResource(ValueProvider.StaticValueProvider.of(outputDir));
+    }
+
+    public Write toResource(ValueProvider<ResourceId> outputDir) {
+      return toBuilder().setOutputDir(outputDir).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<String, byte[]>> input) {
+      checkNotNull(
+          getOutputDir(),
+          "Need to set the output directory of a WholeFileIO.Write transform."
+      );
+
+      input.apply(
+          ParDo.of(
+              new DoFn<KV<String, byte[]>, Void>() {
+                ValueProvider<ResourceId> tmpDir;
+
+                @Setup
+                public void setup() {
+                  tmpDir = ValueProvider.NestedValueProvider.of(
+                      getOutputDir(),
+                      new TemporaryDirectoryBuilder()
+                  );
+                }
+
+                @ProcessElement
+                public void processElement(ProcessContext c) throws IOException {
+                  KV<String, byte[]> kv = c.element();
+
+                  String filename = kv.getKey();
+                  ResourceId tmpFile = tmpDir.get().resolve(filename, RESOLVE_FILE);
+
+                  byte[] bytes = kv.getValue();
+                  try (
+                      OutputStream outStream =
+                          Channels.newOutputStream(FileSystems.create(tmpFile, BINARY))
+                  ) {
+                    outStream.write(bytes);
+                    outStream.flush();
+                  } catch (IOException e) {
+                    LOG.error(
+                        "Failed to write to temporary file [{}] for [{}].",
+                        tmpFile,
+                        getOutputDir().get().resolve(filename, RESOLVE_FILE)
+                    );
+                    FileSystems.delete(
+                        Collections.singletonList(tmpFile),
+                        MoveOptions.StandardMoveOptions.IGNORE_MISSING_FILES
+                    );
+                    throw e;
+                  }
+                }
+
+                @Teardown
+                public void teardown() throws IOException {
+                  try {
+                    // TODO: Replace FileSystems.rename() with a copy and delete to eliminate
+                    // empty directory restriction on Write.
+                    FileSystems.rename(
```

Comment:
Another alternative to this approach is to first delete the existing non-empty destination directory, then rename the source directory to the destination directory. That would be faster than copying all the files over and it would have the effect of removing any existing files and replacing them with the output files.

A fourth (I'm obviously thinking out loud here) is to:

- create the temp dir (in `@Setup`)
- Write to temp file in temp dir (in `@ProcessElement`)
- Rename temp file to destination file (replacing any preexisting file by that filename, even via deletion if necessary) (also in `@ProcessElement`)
- In `@teardown`, the temp directory should actually be empty

This approach would 1) preserve any pre-existing files in the destination directory as long as they don't have the same filename/path as one of the newly written files, 2) write out "finished" files to their intended destination throughout the life of the IO, rather than just all at the end, 3) avoid copying many files (unless of course the `FileSystem` renames via copying, which we can't control).